source/extensions/filters/http/ext_proc/client_impl.h

```diff
@@ -57,6 +57,15 @@ class ExternalProcessorStreamImpl : public ExternalProcessorStream,
     // When the filter object is being destroyed,  `callbacks_` (which is a OptRef to filter object)
     // should be reset to avoid the dangling reference.
     callbacks_.reset();
+
+    // Unregister the watermark callbacks(if any) to prevent access of filter callbacks after
+    // the filter object is destroyed.
+    if (grpc_side_stream_flow_control_ && !watermark_callbacks_removed_) {
+      if (stream_ != nullptr) {
+        stream_.removeWatermarkCallbacks();
+        watermark_callbacks_removed_ = true;
```

Comment:
yeah so as said on the other PR, this smells like we need some refactor. 
why are we removing the callbacks if the stream isn't closed?  can we just close it? can we have a common clean up function?
/wait-any

Reply:
Thanks for the review and suggestion! I just added the [context from another thread](https://github.com/envoyproxy/envoy/pull/35077#discussion_r1686896433) here for future reference.

Per offline discussion, I made the change to use `callback_.has_value()` instead and removed the `watermark_callbacks_removed_ ` flag . I also just verified this new change with internal load test.

PTAL, thanks!